src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -32,73 +33,414 @@
+            "Free region " SIZE_FORMAT ", belonging to %s free partition, must have alloc capacity",
+            idx, partition_name(test_partition));
+    return true;
+  } else {
+    return false;
+  }
+}
+
+inline size_t ShenandoahRegionPartition::leftmost(ShenandoahFreeSetPartitionId which_partition) const {
+  assert (which_partition > NotFree && which_partition < NumPartitions, "selected free partition must be valid");
+  size_t idx = _leftmosts[which_partition];
+  if (idx >= _max) {
+    return _max;
+  } else {
+    assert (in_partition(idx, which_partition), "left-most region must be free");
+    return idx;
+  }
+}
+
+inline size_t ShenandoahRegionPartition::rightmost(ShenandoahFreeSetPartitionId which_partition) const {
+  assert (which_partition > NotFree && which_partition < NumPartitions, "selected free partition must be valid");
+  size_t idx = _rightmosts[which_partition];
+  assert ((_leftmosts[which_partition] == _max) || in_partition(idx, which_partition), "right-most region must be free");
+  return idx;
+}
 
-  assert(_used <= _capacity, "must not use more than we have: used: " SIZE_FORMAT
-         ", capacity: " SIZE_FORMAT ", num_bytes: " SIZE_FORMAT, _used, _capacity, num_bytes);
+inline bool ShenandoahRegionPartition::is_empty(ShenandoahFreeSetPartitionId which_partition) const {
+  assert (which_partition > NotFree && which_partition < NumPartitions, "selected free partition must be valid");
+  return (leftmost(which_partition) > rightmost(which_partition));
 }
 
-bool ShenandoahFreeSet::is_mutator_free(size_t idx) const {
-  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT " (left: " SIZE_FORMAT ", right: " SIZE_FORMAT ")",
-          idx, _max, _mutator_leftmost, _mutator_rightmost);
-  return _mutator_free_bitmap.at(idx);
+size_t ShenandoahRegionPartition::leftmost_empty(ShenandoahFreeSetPartitionId which_partition) {
+  assert (which_partition > NotFree && which_partition < NumPartitions, "selected free partition must be valid");
+  for (size_t idx = _leftmosts_empty[which_partition]; idx < _max; idx++) {
+    if ((membership(idx) == which_partition) && (_free_set->alloc_capacity(idx) == _region_size_bytes)) {
+      _leftmosts_empty[which_partition] = idx;
+      return idx;
+    }
+  }
+  _leftmosts_empty[which_partition] = _max;
+  _rightmosts_empty[which_partition] = 0;
+  return _max;
+}
+
+inline size_t ShenandoahRegionPartition::rightmost_empty(ShenandoahFreeSetPartitionId which_partition) {
+  assert (which_partition > NotFree && which_partition < NumPartitions, "selected free partition must be valid");
+  for (intptr_t idx = _rightmosts_empty[which_partition]; idx >= 0; idx--) {
+    if ((membership(idx) == which_partition) && (_free_set->alloc_capacity(idx) == _region_size_bytes)) {
+      _rightmosts_empty[which_partition] = idx;
+      return idx;
+    }
+  }
+  _leftmosts_empty[which_partition] = _max;
+  _rightmosts_empty[which_partition] = 0;
+  return 0;
```

Comment:
To my earlier comment of using `ssize_t`, that would allow us to signal failure here by returning a -1.

Reply:
In the interest of stability, I'm inclined to leave this convention as is.  Could be persuaded to make the change, but there are probably more than 5 touchpoints that also need to be changed (all invocations, existing documentation, etc.)

Reply:
That sounds reasonable; can be addressed later.